### PR TITLE
BBPBGLIB-1061 SONATA conf: Dont map to internal connectivity

### PR DESCRIPTION
**Pull request by ferdonline**
## Context

Address BBPBGLIB-1061 partially, ensuring, for the moment that we instantiate all projections.

## Scope
 - Consider only some specifically-named edge populations as internal connectivity
 - Other populations shall go as projections

Not all populations should go as projections since circuits still respect order, which is important for the case of e.g. NGV.

## Testing
Existing CI tests

## Review
* [x] PR description is complete
* [x] Coding style (imports, function length, New functions, classes or files) are good
* [ ] Unit/Scientific test added
* [ ] Updated Readme, in-code, developer documentation

### CODE REVIEW

**Review by jorblancoa — DISMISSED** [review]

File: neurodamus/connection_manager.py [hunk 1]
```diff
@@ -322,7 +322,7 @@ def open_synapse_file(self, synapse_file, edge_population, n_files=1, load_offse
 
     # - override if needed
     def _open_synapse_file(self, synapse_file, pop_name, n_nrn_files=None):
-        logging.info("Opening Synapse file %s, population: %s", synapse_file, pop_name)
+        logging.debug("Opening Synapse file %s, population: %s", synapse_file, pop_name)
```
Comment (WeinaJi) [inline neurodamus/connection_manager.py:325]: Why `debug`? I've found this info msg very helpful when I want to know the edges files that neurodamus loads without checking `circuit_config.json`.

Comment (ferdonline) [reply neurodamus/connection_manager.py:325]: I believe very similar information is printed anyway from another side, hence I decreased this specific line to debug. We can still still see it when debugging with `--debug`

Comment (matz-e) [reply neurodamus/connection_manager.py:325]: Who would print that? I agree with @WeinaJi that this is important information that should be in the logs most of the time, and not buried in debug output. There are other things that are less important in the default logs, IIRC.

**Comment by ferdonline** [comment]
@matz-e there's doubt that's useful information, but if you look of a log produced with such run we have
```
[STEP] Circuit NodeB
[INFO] Init SynapseRuleManager. Options: {'load_offsets': False}
[INFO] Loading connections to population: <ConnectionSet: 1761-1761 (NodeB->NodeB)>
[INFO] No matching Connection blocks. Loading all synapses...

|-------20%-------40%-------60%-------80%------100%| [Time taken: 0.00 s]
[INFO]  * [ALL]. Created 2 connections
[STEP] Handling projections...
[INFO] Processing Edge file: /home/runner/work/neurodamus/neurodamus/tests/simulations/usecase3/edges_AB.h5
[INFO]  * NodeA__NodeB__chemical__NodeA-NodeB (Type: Synaptic, Src: NodeA, Dst: NodeB)
[INFO] Loading connections to population: <ConnectionSet: 4058-1761 (NodeA->NodeB)>
```
So for projections already print all information nicely. True I overlooked internal connectivity, which I was trying to get rid of in the first place 🤔
Since we will keep it for the moment, then I guess we are willing to have the duplicated info for projections.

File: neurodamus/connection_manager.py (hunk 1 above)
Comment (ferdonline) [reply neurodamus/connection_manager.py:325]: With new patch:
```
[STEP] Circuit NodeB
[INFO] Processing edge file /home/runner/work/neurodamus/neurodamus/tests/simulations/usecase3/local_edges_B.h5, pop: NodeB__NodeB__chemical
```

**Comment by ferdonline** [comment]
@jorblancoa @matz-e could you take another review pls?

File: neurodamus/io/sonata_config.py [hunk 2]
```diff
@@ -39,6 +41,7 @@ def __init__(self, config_path):
         self._sim_conf = libsonata.SimulationConfig.from_file(config_path)
         self._entries = {}
         self._sections = {}
+        self._internal_edge_pops = set()
```
Comment (matz-e) [inline neurodamus/io/sonata_config.py:44]: Why is this a set if somewhere else we say that order matters?

Do we need this as an additional attribute, too? It seems to me that `circuit_conf["nrnPath"]` has the same contents and could be queried instead?

File: neurodamus/io/sonata_config.py [hunk 3]
```diff
@@ -203,7 +207,7 @@ def Conditions(self):
         conditions["randomize_Gaba_risetime"] = str(conditions["randomize_Gaba_risetime"])
         return {"Conditions": conditions}
 
-    @property
+    @cached_property
     def Circuit(self):
```
[the remark this answers is not in the document]
Comment (ferdonline) [reply neurodamus/io/sonata_config.py:211]: It does unfortunately. That's a key name in BlueConfig

File: neurodamus/io/sonata_config.py (hunk 2 above)
Comment (ferdonline) [reply neurodamus/io/sonata_config.py:44]: This set is internal, to keep track which edge populations are already "exposed" as internal connectivity, so the remaining are exposed as projections. For this purpose we are better served as a set and no need to expose as a property.

File: neurodamus/io/sonata_config.py [hunk 4]
```diff
@@ -56,6 +59,7 @@ def __init__(self, config_path):
 
         self.circuits = libsonata.CircuitConfig.from_file(self.network)
         self._circuit_networks = json.loads(self.circuits.expanded_json)["networks"]
+        logging.debug(self.Circuit)  # compute Circuits so that it also sets _internal_edge_pops
```
[the remark this answers is not in the document]
Comment (ferdonline) [reply neurodamus/io/sonata_config.py:62]: We would need to compute these properties in the `__init__` in the controlled order.. I can't recall why I didn't go that way, maybe it's now easier with the cached property. Let me see

File: neurodamus/io/sonata_config.py (hunk 2 above)
Comment (matz-e) [reply neurodamus/io/sonata_config.py:44]: Yes, but could we not just construct this set on the fly before instantiating all these "projections"?

File: neurodamus/io/sonata_config.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (matz-e) [reply neurodamus/io/sonata_config.py:211]: We're trying to get away from that, though? Do these have to correspond 1-1 with BlueConfig keys?

[the remark this answers is not in the document]
Comment (ferdonline) [reply neurodamus/io/sonata_config.py:211]: We will eventually. This Sonata reader is basically a conversion layer, so we expose keys like those in BlueConfigs.

File: neurodamus/io/sonata_config.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ferdonline) [reply neurodamus/io/sonata_config.py:62]: I missed I already do that in `__init__`. Let me make this cleaner then

File: neurodamus/io/sonata_config.py (hunk 2 above)
Comment (ferdonline) [reply neurodamus/io/sonata_config.py:44]: We have `Circuit` and `Projection` properties which both consume edges. Hence this change needs a shared object

**Review by jorblancoa — APPROVED** [review]

**Pull request merged**